fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py

```diff
@@ -0,0 +1,243 @@
+
+    def get_min_max_sentinels(
+        self,
+        sentinel_prefix: Optional[str] = "<SENTINEL_ID",
+        integer_find_regex: Optional[str] = "\d{1,}",
+    ) -> Tuple[int, int]:
+        """
+        returns a Tuple [min encountered sentinel name, max encountered sentinel name]
+
+        For example, if the vocab contains:
+
+        SENTINEL_ID_101: 1000,
+        SENTINEL_ID_102: 1001,
+        SENTINEL_ID_103: 1002,
+
+        will return [101,103]
+        """
+        min_token = None
+        max_token = None
+
+        for k, _ in self._tokenizer.get_added_vocab().items():
+            if sentinel_prefix in k:
+                val = re.findall(integer_find_regex, k)
+                if len(val) != 1:
+                    raise Exception(f"expected exactly one integer number in {k} but found {val}")
+                val = val[0]
+                val = int(val)
+
+                if (min_token is None) or (val < min_token):
+                    min_token = val
+
+                if (max_token is None) or (val > max_token):
+                    max_token = val
+
+        if (min_token is None) or (max_token is None):
+            raise Exception(f'Could not find any sentinels with the prefix "{sentinel_prefix}"')
+
+        return (min_token, max_token)
+
+    def get_token_id(self, token_str: str) -> int:
+        ans = self._tokenizer.token_to_id(token_str)
+        assert ans is not None, f"could not find token id for token:{token_str}!"
+        return ans
+
+    def __call__(
+        self,
+        sample_dict: NDict,
+        key_in: str,
+        key_out_tokenized_object: Optional[str] = None,
+        key_out_tokens_ids: Optional[str] = None,
+        key_out_attention_mask: Optional[str] = None,
+        convert_attention_mask_to_bool: Optional[bool] = True,
+    ) -> NDict:
+        # if self._verbose:
+        #     print(
+        #         f'PID:{os.getpid()} FastModularTokenizer op sample_id {sample_dict["data.sample_id"]} key_in={key_in} pdb={sample_dict["pdb"]} HeavyChain: {sample_dict["Hchain"]} LightChain: {sample_dict["Lchain"]}'
+        #     )
+
+        data_lst = sample_dict[key_in]
+        if not isinstance(data_lst, list):
+            # data_lst is a list of named tuples of type collections.namedtuple("TypedInput", ["input_type", "input_string", "max_len"])
+            raise Exception(
+                f"Expected key_in={key_in} to point to a list of inputs, and instead got a {type(data_lst)}. value={data_lst}"
+            )
+
+        if self._validate_ends_with_eos is not None:
+            if not data_lst[-1].input_string.rstrip().endswith(self._validate_ends_with_eos):
+                raise Exception(
+                    f"self._validate_ends_with_eos was set to {self._validate_ends_with_eos}, but about to encode a string that does not end with it. The str end was: {data_lst[-1].input_string}"
+                )
+
+        encoded = self._tokenizer.encode_list(data_lst)
```

Comment:
What are the differences from FastTokenizer op?
Isn't just this line and lines 174-179?

Reply:
That's mostly it. There are several more changes - I no longer use get_vocab(), (since it is not well-defined for modulartokenizer) which is used in several places in FastTokenizer ops (finding sentinel IDs, max token IDS, etc. - I implemented these using either get_special_vocab, which is well-defined, or built-in tokenizer functions)